### PR TITLE
Simplified Preprocessor and Postprocessor Usage and Defaults

### LINKED ISSUE #708: Simpler usage for PreprocessorPyTorch
**Is your feature request related to a problem? Please describe.**
To create a very pytorch-specific preprocessor, you have to write a lot of code (some of which includes numpy and some of which includes pytorch). Here is an example of truncating a video:
```
from art.defences.preprocessor.preprocessor import PreprocessorPyTorch                   
                                                                                         
class VideoTruncate(PreprocessorPyTorch):                                                
    """                                                                                  
    Simple PyTorch preprocessing (truncate frames from video)                            
    """                                                                                  
                                                                                         
    def __init__(self, max_frames=64, device_type="gpu"):                                
        import torch                                                                     
        super().__int__()                                                                
        max_frames = int(max_frames)                                                     
        if max_frames < 0:                                                               
            raise ValueError(f"max_frames {max_frames} cannot be negative")              
        self.max_frames = max_frames                                                     
        if device_type == "cpu" or not torch.cuda.is_available():                        
            self._device = torch.device("cpu")                                           
        else:                                                                            
            cuda_idx = torch.cuda.current_device()                                       
            self._device = torch.device("cuda:{}".format(cuda_idx))                      
                                                                                         
    def forward(self, x, y=None):                                                        
        """                                                                              
        Assuming video shape (batch_size, frames, height, width, channels)               
        """                                                                              
        return x[:, :self.max_frames], y                                                 
                                                                                         
    def estimate_forward(self, x, y=None):                                               
        return self.forward(x, y=y)

    def __call__(self, x, y=None):                                                       
        x = torch.tensor(x, device=self._device)                                         
        if y is not None:                                                                
            y = torch.tensor(y, device=self._device)                                     
        with torch.no_grad():                                                            
            x, y = self.forward(x, y)                                                    
                                                                                         
        x = x.cpu().numpy()                                                              
        if y is not None:                                                                
            y = y.cpu().numpy()                                                          
                                                                                         
        return x, y                                                                      
                                                                                         
    def estimate_gradient(self, x, gradients):                                           
        x = torch.tensor(x, device=self._device, requires_grad=True)                     
        gradients = torch.tensor(gradients, device=self._device)                         
        x_orig = x                                                                       
        x = self.estimate_forward(x)                                                     
        x.backward(gradients)                                                            
        gradients = x_orig.grad.detach().cpu().numpy()                                   
        return gradients                                                                 
                                                                                         
    def apply_fit(self):                                                                 
        return True                                                                      
                                                                                         
    def apply_predict(self):                                                             
        return True                                                                      
                                                                                         
    def fit(self):                                                                       
        pass
```

**Describe the solution you'd like**
I would like to be able to write the following code instead (only the forward pass should be required), which would be very intuitive for those who know pytorch:
```
class VideoTruncate(PreprocessorPyTorch):                                                
    """                                                                                  
    Simple PyTorch preprocessing (truncate frames from video)                            
    """                                                                                  
                                                                                         
    def __init__(self, max_frames=64):                                                   
        super().__int__()                                                                
        max_frames = int(max_frames)                                                     
        if max_frames < 0:                                                               
            raise ValueError(f"max_frames {max_frames} cannot be negative")              
        self.max_frames = max_frames                                                     
                                                                                         
    def forward(self, x, y=None):                                                        
        """                                                                              
        Assuming video shape (batch_size, frames, height, width, channels)               
        """                                                                              
        return x[:, :self.max_frames], y
```

This requires modifications to `PreprocessorPyTorch`. 

Ideally, this would also change `PyTorchEstimator ` to work consistently when preprocessing defenses has a list of size 1. Similar modifications would be helpful on the TensorFlow side as well.

**Describe alternatives you've considered**
None. Currently, things work, but are rather clunky and require more deeply understanding the subclasses. With this approach, you could reduce the user knowledge and effort required, and ensure greater consistency for pytorch and numpy interfaces. 

**Additional context**
If this is of interest, I am happy to contribute. Just let me know, and point me to the milestone you would want it in.

### PR DESCRIPTION
**Pull request by davidslater**
# Description

Please include a summary of the change, motivation and which issue is fixed. Any dependencies changes should also be included.

Fixes #708

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by davidslater** [comment]
@beat-buesser I would be interested in getting your take on this and its usage on the spatial smoothing defense before I apply similar edits to all the preprocessor defenses in this PR.

This should make it much simpler for pytorch and tensorflow users to create preprocessors, since they only need to produce init and forward functions (and estimate_forward if not differentiable).

**Comment by davidslater** [comment]
The only other default I was considering was setting `estimate_gradient` in the base `Preprocessor` class:

```
    def estimate_gradient(self, x: np.ndarray, grad: np.ndarray) -> np.ndarray:          
        return grad
```

But I decided against it. Thoughts?

**Comment by beat-buesser** [comment]
I agree on deciding against moving `estimate_gradient` to `Preprocessor`, I think the main reason is that this would be the place where a Backward Pass Differentiable Approximation would be implemented for a Numpy implementation. Most of the current defences' `estimate_gradient` methods look similar because they are all using Identity as approximation.

Let me take a closer look at the other changes.

**Comment by beat-buesser** [comment]
I just realised you probably meant above to have `estimate_gradient` for BPDA with Identity in the `Preprocessor` base as default and only implement `estimate_gradient` in a defence if it has a BPDA better than Identity or accurate gradients? wHat was your reason to decide against it?

The changes so far look good to me, the changed preprocessing defences changed look very nice and clean.

**Comment by davidslater** [comment]
Yes, that's what I meant. I was on the fence about adding it - but since most of the defenses use identity for BPDA, I think that it makes sense. I'll add it if you're good with it.

Aside from that, I'll make updates across the preprocessing defences as well as `art/preprocessing/standardisation_mean_std/*`.

Also, is there a reason that `art/preprocessing/preprocessing.py` exists? It's an empty file.

**Comment by davidslater** [comment]
Also, while I'm at it, do you think I should add similar updates to the Postprocessor and related classes?

**Comment by beat-buesser** [comment]
I agree, moving a default `estimate_gradient` to `Preprocessor` makes sense. Maybe we can add a comment to the docstring of `Preprocessor` to alert users that the default is BPDA with Identity.

I agree on updating `art.preprocessing.*`.

About `art/preprocessing/preprocessing.py`, we were discussing if it would make sense to move the Preprocessor API to `art.preprocessing` into `art/preprocessing/preprocessing.py` and reuse it in `art.defences.preprocessor`. It could make sense if we expect more general preprocessing tools than defense preprocessing tools in the future with Expectation over Transformation and environment simulators coming to `art.preprocessing`.

I think updating Postprocessor similar to Preprocessor makes would be great, if you have time, to keep noth APIs as close as possible.

**Comment by davidslater** [comment]
I updated postprocessor. Let me know if there are other things you would like me to do / test.

Regarding the placement of preprocessor, I don't have an opinion, so long as the changes are backwards compatible.

**Review by beat-buesser — APPROVED** [review]
Hi @davidslater Thank you very much for simplifying the Preprocessor and Postprocessor defences and their APIs! This will greatly facilitate the integration of new defences.
